nrf52-hal-common/src/spim.rs

```diff
@@ -201,100 +241,123 @@ impl<T> Spim<T> where T: SpimExt {
             // The PTR field is a full 32 bits wide and accepts the full range
             // of values.
-            unsafe { w.ptr().bits(tx_buffer.as_ptr() as u32) }
-        );
-        self.0.txd.maxcnt.write(|w|
+
+            // tx_data_ptr:
+            tx_buffer.as_ptr() as u32,
+
             // We're giving it the length of the buffer, so no danger of
             // accessing invalid memory. We have verified that the length of the
-            // buffer fits in an `u8`, so the cast to `u8` is also fine.
+            // buffer fits in an `u8`, so the cast to the type of maxcnt
+            // is also fine.
             //
-            // The MAXCNT field is 8 bits wide and accepts the full range of
-            // values.
-            unsafe { w.maxcnt().bits(tx_buffer.len() as _) }
-        );
+            // Note that that nrf52840 maxcnt is a wider
+            // type than a u8, so we use a `_` cast rather than a `u8` cast.
+            // The MAXCNT field is thus at least 8 bits wide and accepts the full
+            // range of values that fit in a `u8`.
 
-        // Tell the RXD channel it doesn't need to read anything
-        self.0.rxd.maxcnt.write(|w|
-            // This is safe for the same reasons that writing to TXD.MAXCNT is
-            // safe. Please refer to the explanation there.
-            unsafe { w.maxcnt().bits(0) }
-        );
+            // tx_len:
+            tx_buffer.len() as _,
 
-        // Start SPI transaction
-        chip_select.set_low();
-        self.0.tasks_start.write(|w|
-            // `1` is a valid value to write to task registers.
-            unsafe { w.bits(1) }
-        );
+            // Set up the DMA read
+            // This is safe for the same reasons that writing to TXD.PTR is
+            // safe. Please refer to the explanation there.
 
-        // Wait for transmission to end
-        while self.0.events_end.read().bits() == 0 {}
+            // rx_data_ptr:
+            rx_buffer.as_mut_ptr() as u32,
+            // This is safe for the same reasons that writing to TXD.MAXCNT is
+            // safe. Please refer to the explanation there.
 
-        // Reset the event, otherwise it will always read `1` from now on.
-        self.0.events_end.write(|w| w);
+            // rx_len:
+            rx_buffer.len() as _,
+            // chip select callback
+            |cs|{if cs {chip_select.set_low()} else {chip_select.set_high()} }
+        )
+    }
 
-        // Conservative compiler fence to prevent optimizations that do not
-        // take in to account actions by DMA. The fence has been placed here,
-        // after all possible DMA actions have completed
-        compiler_fence(SeqCst);
+    /// Write to an SPI slave
+    ///
+    /// This method uses the provided chip select pin to initiate the
+    /// transaction, then transmits all bytes in `tx_buffer`.
+    ///
 
-        // End SPI transaction
-        chip_select.set_high();
+    pub fn write(&mut self,
```

Comment:
This  old method now have the same name as write in the embedded_hal trait, it causes some problem for a user of the library.

It is still works -- it is possible to choose which function to use and the number of parameters are not the same so the user will notice.

Reply:
I'm inclined to accept this for now, so we can get this PR merged. But yes, it would be better if there weren't any name collisions.